File: core/src/androidMain/kotlin/ru/mobileup/template/core/map/internal/LogoController.kt
package ru.mobileup.template.core.map.internal

import com.yandex.mapkit.logo.HorizontalAlignment
import com.yandex.mapkit.logo.Padding
import com.yandex.mapkit.logo.VerticalAlignment
import com.yandex.mapkit.mapview.MapView
import com.yandex.mapkit.logo.Alignment as LogoAlignment

internal class LogoController(
    private val mapView: MapView
) {
    fun setPosition(position: MapLogoPosition) {
        mapView.mapWindow.map.logo.setAlignment(
            LogoAlignment(
                position.horizontalAlignment.toYandexAlignment(),
                position.verticalAlignment.toYandexAlignment()
            )
        )
        mapView.mapWindow.map.logo.setPadding(
            Padding(
                position.horizontalPaddingPx,
                position.verticalPaddingPx
            )
        )
    }

    private fun MapLogoHorizontalAlignment.toYandexAlignment(): HorizontalAlignment {
        return when (this) {
            MapLogoHorizontalAlignment.Left -> HorizontalAlignment.LEFT
            MapLogoHorizontalAlignment.Right -> HorizontalAlignment.RIGHT
        }
    }

    private fun MapLogoVerticalAlignment.toYandexAlignment(): VerticalAlignment {
        return when (this) {
            MapLogoVerticalAlignment.Top -> VerticalAlignment.TOP
            MapLogoVerticalAlignment.Bottom -> VerticalAlignment.BOTTOM
        }
    }
}

File: iosApp/iosApp/Source/Map/Subcontrollers/LogoController.swift
import YandexMapsMobile
import Shared

final class LogoController {

    private let mapView: YMKMapView

    init(mapView: YMKMapView) {
        self.mapView = mapView
    }

    func setPosition(position: MapLogoPosition) {
        let logo = mapView.mapWindow.map.logo
        logo.setAlignmentWith(
            YMKLogoAlignment(
                horizontalAlignment: position.horizontalAlignment.ymkAlignment,
                verticalAlignment: position.verticalAlignment.ymkAlignment
            )
        )
        logo.setPaddingWith(
            YMKLogoPadding(
                horizontalPadding: UInt(max(0, position.horizontalPaddingPx)),
                verticalPadding: UInt(max(0, position.verticalPaddingPx))
            )
        )
    }
}

private extension MapLogoHorizontalAlignment {
    var ymkAlignment: YMKLogoHorizontalAlignment {
        // Kotlin enums bridge to objects, so compare by entry name rather than switch.
        self == MapLogoHorizontalAlignment.left ? .left : .right
    }
}

private extension MapLogoVerticalAlignment {
    var ymkAlignment: YMKLogoVerticalAlignment {
        self == MapLogoVerticalAlignment.top ? .top : .bottom
    }
}
